adr/0007-package-variables-files-revisit.md

```diff
@@ -0,0 +1,135 @@
+
+For the `default` / `prompt` cases, below is an example of the current behavior with the under the hood values annotated:
+
+```yaml
+variables:
+  - name: "NOTHING"
+    # default: ""   <- currently under the hood
+    # prompt: false <- currently under the hood
+  - name: "NO_PROMPT"
+    default: "no-prompt"
+    # prompt: false <- currently under the hood
+  - name: "NO_DEFAULT"
+    # default: ""   <- currently under the hood
+    prompt: true
+  - name: "BOTH"
+    default: "both"
+    prompt: true
+```
+
+```shell
+zarf package deploy package.zst --confirm --set NOTHING=hello --set NO_PROMPT=world --set NO_DEFAULT=today --set BOTH=tommorow
+# This currently works, setting NOTHING to "hello", NO_PROMPT to "world", NO_DEFAULT to "today" and BOTH to "tomorrow"
+
+zarf package deploy package.zst --confirm --set NOTHING=hello --set NO_DEFAULT=today
+# This currently works, setting NOTHING to "hello", NO_PROMPT to "no-prompt", NO_DEFAULT to "today" and BOTH to "both"
+
+zarf package deploy package.zst --confirm --set NOTHING=hello
+# This currently works, setting NOTHING to "hello", NO_PROMPT to "no-prompt", NO_DEFAULT to "" and BOTH to "both"
+
+zarf package deploy package.zst --confirm
+# This currently works, setting NOTHING to "", NO_PROMPT to "no-prompt", NO_DEFAULT to "" and BOTH to "both"
+
+zarf package deploy package.zst
+# This currently works, setting NOTHING to "" and NO_PROMPT to "no-prompt" and prompting the user for NO_DEFAULT and BOTH
+```
+
+This introduces confusion due to two separate issues/questions:
+
+1. What if a package author doesn't want a `default` set?
+1. What is the resulting behavior of `--confirm` and `prompt` in combination?
+
+## Options
+
+With the above context below are options to address the concerns for both `prompt` and `default` keeping in mind that file variables will be added.
+
+### Default
+
+The default confusion could be addressed in one of the following ways:
+
+1. Remove the `default` key altogether and rely on package authors to create example overrides files.
+   - *Pros*: This makes things simpler and allows us not to have to worry about abuses of this key.
+   - *Cons*: This removes options from package authors where they may want to set a legitimate default.  Overrides files also don't natively travel along with the package.
+1. Allow the `default` key to be unset if it does not exist (no implicit "")
+   - *Pros*: This allows package authors to set defaults that travel inside of a package and can be potentially carried into generated example override files.
+   - *Cons*: This would require the use of a pointer in the package types definitions.
+1. Keep the current `default` behavior and force an implicit ""
+   - *Pros*: This is not a breaking change and allows us to not have to have a pointer in our type definitions
+   - *Cons*: This forces an implicit default onto package creators.
+
+### Prompt
```

Comment:
Want to quickly through out another option where we essentially turn `prompt` into an 'interactive only' configuration and ignore it during `--confirm` but then we add another field to the schema along the lines of `requireInput` where we would error out if someone tried to deploy with `--confirm` without setting that field.

- *Pros*: Non-breaking change, avoids NEEDing a pointer (if that's an issue), gives package creates even more control, not too different from what we currently do.
- *Cons*: Another addition to the schema, having `prompt` only really matter in 'interactive mode' feels weird at first

Reply:
My main concern with this ADR is I think package creators should have the ability ensure a field is set during deploy without making deploying more cumbersome for the user (requiring every field be set, adding a new flag for using defaults, etc.). I am not too concerned with an implicit empty string if the package creator can dictate that specific variables are not allowed to be empty strings.

Reply:
> I am not too concerned with an implicit empty string if the package creator can dictate that specific variables are not allowed to be empty strings.

My problem with this is it would expect that the author will know there will be an implicit empty string, and forces them to add the the regex matcher to avoid unintentional side effects. It will be both error prone and annoying to the author.

Reply:
Yeah and I see the `prompt` and `default` keys as separate (but related) issues.  I am more firm on `default` being unset by default but could also support the addition of another `--ignore-prompt` flag or another schema key of `requireInput` (personally would lean toward another flag in that case).